upring.js

```diff
@@ -257,8 +257,69 @@ UpRing.prototype.request = function (obj, callback, _count) {
 
 UpRing.prototype.requestp = promisify(UpRing.prototype.request)
 
-function retry (that, obj, callback, _count) {
-  that.request(obj, callback, _count)
+UpRing.prototype.fire = function (obj, callback, _count) {
+  callback = callback || fireCallback
+  if (!this.isReady) {
+    this.once('up', this.fire.bind(this, obj, callback))
+    return
+  }
+
+  if (this._hashring.allocatedToMe(obj.key)) {
+    this.logger.trace({ msg: obj }, 'local call')
+    this._dispatch(obj, dezalgo(callback.bind(this)))
+  } else {
+    let peer = this._hashring.lookup(obj.key)
+    this.logger.trace({ msg: obj, peer }, 'remote call')
+
+    let upring = peer.meta.upring
+    if (!upring || !upring.address || !upring.port) {
+      callback.call(this, new Error('peer has invalid upring metadata'))
+      return
+    }
+
+    // TODO simplify all this logic
+    // and avoid allocating a closure
+    if (typeof _count !== 'number') {
+      _count = 0
+    } else if (_count === 3) {
+      callback.call(this, new Error('retried three times'))
+      return
+    } else {
+      _count++
+    }
+
+    const conn = this.peerConn(peer)
+
+    if (conn.destroyed) {
+      // TODO make this dependent on the gossip interval
+      setTimeout(retry, 500, this, 'fire', obj, callback, _count)
+      return
+    }
+
+    conn.fire(obj, (err) => {
+      if (err) {
+        // the peer has changed
+        if (this._hashring.lookup(obj.key).id !== peer.id || conn.destroyed) {
+          // TODO make this dependent on the gossip interval
+          setTimeout(retry, 500, this, 'fire', obj, callback, _count)
+          return
+        }
+      }
+      callback.call(this, err)
```

Comment:
To support [this](https://github.com/delvedor/upring/blob/d30b1569bb852477ff678dfebc7f569ecf183247/upring.js#L315-L319).